### ISSUE TITLE
Executor becomes unresponsive with too many open requests waiting for deleted function

### ISSUE THREAD

**Issue by erwinvaneyk** [fission/fission#887]
The fission executor seems to become unresponsive if there are (many?) open `getServiceForFunction` requests for a deleted function. 

The following logs are basically repeating every second: (fut is the function)
```
2018/09/07 08:07:02 [fut] Deleting cache entry for invalid address : fut-vngggrhw.fission-function
2018/09/07 08:07:02 Waiting for concurrent request for the same function: &ObjectMeta{Name:fut,GenerateName:,Namespace:default,SelfLink:/apis/fission.io/v1/namespaces/default/functions/fut,UID:cbe3d4fb-b25d-11e8-b8f0-42010a9a0046,ResourceVersion:46903,Generation:0,CreationTimestamp:2018-09-07 05:20:55 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
2018/09/07 08:07:02 Error validating service address for function fut: services "fut-vngggrhw" not found
2018/09/07 08:07:02 [fut] Deleting cache entry for invalid address : fut-vngggrhw.fission-function
2018/09/07 08:07:02 Waiting for concurrent request for the same function: &ObjectMeta{Name:fut,GenerateName:,Namespace:default,SelfLink:/apis/fission.io/v1/namespaces/default/functions/fut,UID:cbe3d4fb-b25d-11e8-b8f0-42010a9a0046,ResourceVersion:46903,Generation:0,CreationTimestamp:2018-09-07 05:20:55 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
2018/09/07 08:07:03 Error validating service address for function fut: services "fut-vngggrhw" not found
2018/09/07 08:07:03 [fut] Deleting cache entry for invalid address : fut-vngggrhw.fission-function
2018/09/07 08:07:03 Waiting for concurrent request for the same function: &ObjectMeta{Name:fut,GenerateName:,Namespace:default,SelfLink:/apis/fission.io/v1/namespaces/default/functions/fut,UID:cbe3d4fb-b25d-11e8-b8f0-42010a9a0046,ResourceVersion:46903,Generation:0,CreationTimestamp:2018-09-07 05:20:55 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
2018/09/07 08:07:03 Error validating service address for function fut: services "fut-vngggrhw" not found
2018/09/07 08:07:03 [fut] Deleting cache entry for invalid address : fut-vngggrhw.fission-function
2018/09/07 08:07:03 Waiting for concurrent request for the same function: &ObjectMeta{Name:fut,GenerateName:,Namespace:default,SelfLink:/apis/fission.io/v1/namespaces/default/functions/fut,UID:cbe3d4fb-b25d-11e8-b8f0-42010a9a0046,ResourceVersion:46903,Generation:0,CreationTimestamp:2018-09-07 05:20:55 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
```

Finally, after a while (probably timeout related) it finally cancels the requests.
```
10.36.1.3 - - [07/Sep/2018:06:46:19 +0000] "POST /v2/getServiceForFunction HTTP/1.1" 500 141
10.36.1.3 - - [07/Sep/2018:06:46:19 +0000] "POST /v2/getServiceForFunction HTTP/1.1" 500 141
10.36.1.3 - - [07/Sep/2018:06:46:19 +0000] "POST /v2/getServiceForFunction HTTP/1.1" 500 141
10.36.1.3 - - [07/Sep/2018:06:46:19 +0000] "POST /v2/getServiceForFunction HTTP/1.1" 500 141
10.36.1.3 - - [07/Sep/2018:06:46:19 +0000] "POST /v2/getServiceForFunction HTTP/1.1" 500 141
10.36.1.3 - - [07/Sep/2018:06:46:19 +0000] "POST /v2/getServiceForFunction HTTP/1.1" 500 141
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
2018/09/07 08:18:28 Error: 500: Error getting service for function fut in namespace default: Resource not found - key 'cbe3d4fb-b25d-11e8-b8f0-42010a9a0046_46903' not found
```

Not only the logging is verbose (possibly also slowing down the executor further), the executor seems to become unresponsive to any future function CRD changes for as long as it i; new functions are no longer deployed.

**To reproduce**

1. Create a function 
2. Invoke function at high-concurrency (bug manifested at 1000 requests/second) 
3. While requests are in-flight, delete the function

tl;dr I think this bug is two-fold:
- `getServiceForFunction` requests are not canceled immediately if the targeted function has been deleted.
- At too many open requests, the executor becomes unresponsive to any other updated function CRDs.

Although you could argue that the system simply got overloaded, instead of the current situation we should: fail fast, and not impact future interactions with the system.

**Comment by soamvasani** [comment]
@life1347 Can you take a look at this? Your router optimization definitely helps some aspect of this -- are there other issues?

**Comment by life1347** [comment]
It's a known issue when I tried to optimize the router performance.

The root cause is router sends too many requests when no corresponding function service entry in the cache and caused executor overloaded. 

To solve this problem we have to limit the maximum amounts of `getServiceForFunction` to executor per function. It will increase router performance under high concurrency situation. 

And agree with the idea `fail fast`. I will make sure it was fixed in latest router optimization.

**Comment by thenamly** [comment]
We have been hitting similar problem (#707) around `getServiceForFunction` in last 6 months. Router optimization (#856) decreases the error rate but not completely. After all workarounds, scaling problems of fission workflows mostly come from fission.
